src/OpenApi/src/Extensions/OpenApiDocumentExtensions.cs

```diff
@@ -14,28 +14,33 @@ internal static class OpenApiDocumentExtensions
     /// <param name="document">The <see cref="OpenApiDocument"/> to register the schema onto.</param>
     /// <param name="schemaId">The ID that serves as the key for the schema in the schema store.</param>
     /// <param name="schema">The <see cref="IOpenApiSchema" /> to register into the document.</param>
-    /// <returns>An <see cref="IOpenApiSchema"/> with a reference to the stored schema.</returns>
-    public static IOpenApiSchema AddOpenApiSchemaByReference(this OpenApiDocument document, string schemaId, IOpenApiSchema schema)
+    /// <param name="schemaReference">An <see cref="IOpenApiSchema"/> with a reference to the stored schema.</param>
+    /// <returns>Whether the schema was added or already existed</returns>
+    public static bool AddOpenApiSchemaByReference(this OpenApiDocument document, string schemaId, IOpenApiSchema schema, out OpenApiSchemaReference schemaReference)
     {
-        document.Components ??= new();
-        document.Components.Schemas ??= new Dictionary<string, IOpenApiSchema>();
-        document.Components.Schemas[schemaId] = schema;
+        // Make sure the document has a workspace,
+        // AddComponent will add it to the workspace when adding the component.
         document.Workspace ??= new();
-        var location = document.BaseUri + "/components/schemas/" + schemaId;
-        document.Workspace.RegisterComponentForDocument(document, schema, location);
+        // AddComponent will only add the schema if it doesn't already exist.
+        var schemaAdded = document.AddComponent(schemaId, schema);
 
         object? description = null;
         object? example = null;
+        object? defaultAnnotation = null;
         if (schema is OpenApiSchema actualSchema)
```

Comment:
Could add a null check for metadata here, then the next three lines don't need the `?`.